packages/ui/src/composables/usePopover.ts

```diff
@@ -0,0 +1,196 @@
+export type Offset = number | [number, number]
+
+type Coords = { x: number, y: number }
+type AlignCoords = { main: number, cross: number }
+
+const coordsToCss = ({ x, y }: Coords) => ({ left: `${x}px`, top: `${y}px` })
+
+const parsePlacement = (placement: Placement) => {
+  const [position, align] = placement.split('-') as [PlacementPosition, PlacementAlignment | undefined]
+
+  return { position, align: align || 'center' }
+}
+
+const parseOffset = (offset: Offset): AlignCoords => {
+  return Array.isArray(offset) ? { main: offset[0], cross: offset[1] } : { main: offset, cross: 0 }
+}
+
+const calculateContentAlignment = (align: PlacementAlignment, anchorStart: number, anchorSize: number, contentSize: number) => {
+  if (align === 'start') { return anchorStart }
+  if (align === 'end') { return anchorStart + anchorSize - contentSize }
+
+  return anchorStart + (anchorSize - contentSize) / 2
+}
+
+const calculateContentCoords = (placement: Placement, anchor: DOMRect, content: DOMRect) => {
+  const { position, align } = parsePlacement(placement)
+
+  if (position === 'top') {
+    return {
+      x: calculateContentAlignment(align, anchor.left, anchor.width, content.width),
+      y: anchor.top - content.height,
+    }
+  }
+  if (position === 'left') {
+    return {
+      y: calculateContentAlignment(align, anchor.top, anchor.height, content.height),
+      x: anchor.left - content.width,
+    }
+  }
+  if (position === 'right') {
+    return {
+      y: calculateContentAlignment(align, anchor.top, anchor.height, content.height),
+      x: anchor.right,
+    }
+  }
+
+  // if position === 'bottom'
+  return {
+    x: calculateContentAlignment(align, anchor.left, anchor.width, content.width),
+    y: anchor.bottom,
+  }
+}
+
+const calculateOffsetCoords = (placement: Placement, offset: Offset): Coords => {
+  const { position } = parsePlacement(placement)
+  const { main, cross } = parseOffset(offset)
+
+  if (position === 'top') {
+    return { y: -main, x: cross }
+  }
+
+  if (position === 'right') {
+    return { y: cross, x: main }
+  }
+
+  if (position === 'left') {
+    return { y: cross, x: -main }
+  }
+
+  return { y: main, x: cross }
+}
+
+/** Returns how much content overflow */
+const calculateContentOverflow = (coords: Coords, content: DOMRect, root: DOMRect) => {
+  const xMax = root.right
+  const yMax = root.bottom
+  const xMin = root.left
+  const yMin = root.top
+
+  return {
+    top: Math.max(yMin - coords.y, 0),
+    bottom: Math.max((coords.y + content.height) - yMax, 0),
+    left: Math.max(xMin - coords.x, 0),
+    right: Math.max((coords.x + content.width) - xMax, 0),
+  }
+}
+
+const clamp = (min: number, v: number, max: number) => Math.max(Math.min(v, max), min)
+
+const calculateClipToEdge = (coords: Coords, content: DOMRect, anchor: DOMRect, root: DOMRect) => {
+  const { top, bottom, left, right } = calculateContentOverflow(coords, content, root)
```

Comment:
I'd rather not use destructuring here since those names are used very commonly and can be misinterpreted in below expression in a quick glance.